v2/apiserver/internal/core/kubernetes/logs_store.go

```diff
@@ -0,0 +1,141 @@
+package kubernetes
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/brigadecore/brigade/v2/apiserver/internal/core"
+	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+// logsStore is a Kubernetes-based implementation of the core.LogsStore
+// interface.
+type logsStore struct {
+	kubeClient kubernetes.Interface
+}
+
+// NewLogsStore returns a Kubernetes-based implementation of the core.LogsStore
+// interface. It can stream logs directly from a Worker or Job's underlying pod.
+// In practice, this is useful for very near-term retrieval of Worker and Job
+// logs without incurring the latency inherent in other implementations that
+// rely on a log aggregator having forwarded and stored log entries. This
+// implementation will error, however, once the relevant pod has been deleted.
+// Callers should be prepared to fall back on another implementation of the
+// core.LogsStore interface, with the assumption that by the time a Worker's or
+// Job's pod has been deleted, all of its logs have been aggregated and stored.
+func NewLogsStore(kubeClient kubernetes.Interface) core.LogsStore {
+	return &logsStore{
+		kubeClient: kubeClient,
+	}
+}
+
+func (l *logsStore) StreamLogs(
+	ctx context.Context,
+	project core.Project,
+	event core.Event,
+	selector core.LogsSelector,
+	opts core.LogStreamOptions,
+) (<-chan core.LogEntry, error) {
+	podName, containerName := criteriaFromSelector(event.ID, selector)
+
+	req := l.kubeClient.CoreV1().Pods(project.Kubernetes.Namespace).GetLogs(
+		podName,
+		&v1.PodLogOptions{
+			Container:  containerName,
+			Timestamps: true,
+		},
+	)
+	podLogs, err := req.Stream(ctx)
+	if err != nil {
+		return nil, errors.Wrapf(
```

Comment:
Great question. There probably _is_ a specific error we can trap that would point to the pod we were tying to get logs from no longer existing. I'll have to play around a bit and see if I can figure out what that error looks like.

But the second part of your question is good, too. Does it _matter_ what the error was as long as the caller is always prepared to fall back on the second approach?

This admittedly needs some more thought.

I suppose on the one hand, always falling back regardless of what the error was seems like a good road to recovering from the error. On the other hand, suppose the error was for some reason _other than_ the pod no longer existing. If we fall back on the second approach, we're possibly just masking a more serious problem.

I'm leaning toward agreeing that we should try to spot errors that are specifically the result of the pod no longer existing and return an error that clearly indicates to the caller that falling back on the second strategy is worthwhile.

Reply:
> I'm leaning toward agreeing that we should try to spot errors that are specifically the result of the pod no longer existing and return an error that clearly indicates to the caller that falling back on the second strategy is worthwhile.

Yes, this sounds reasonable to me as well.  It can definitely be created a separate ticket/follow-up, if preferred.

Reply:
It's ok. I think it's worth fixing now.

Reply:
I fixed this. I have the warm logs implementation (k8s-based) checking whether it gets back a 404 from the k8s apiserver when it goes looking for pod logs. If it does, it translates that into a `*meta.ErrNotFound`. The service layer falls back on cool logs when it sees that specific error. All _other_ errors coming from warm logs will _not_ cause the service layer to silently fall back on the cool logs.

Reply:
Awesome ^^ This sounds great.